Sources/WebSocketKit/PMCE.swift

```diff
@@ -0,0 +1,753 @@
+            try compressorBox.value?.finishStream()
+        }
+        catch {
+            logger.error("PMCE:error finishing stream(s) : \(error)")
+        }
+        
+        do {
+            logger.debug("PMCE: stopping decompressor stream...")
+            try decompressorBox.value?.finishStream()
+        }
+        catch {
+            logger.error("PMCE:error finishing stream(s) : \(error)")
+        }
+        
+
+    }
+
+    /// websocket send calls this to compress.
+    public func compressed(_ buffer: ByteBuffer,
+                            fin: Bool = true,
+                            opCode: WebSocketOpcode = .binary) throws -> WebSocketFrame {
+        
+        guard let channel = channel else {
+            throw IOError(errnoCode: 0, reason: "PMCE: channel not configured.")
+        }
+        let notakeover = !shouldTakeOverContext()
+
+        let startSize = buffer.readableBytes
+        
+        if logging {
+            logger.trace("compressing \(startSize) bytes for \(opCode)")
+        }
+        do {
+            var mutBuffer = buffer
+
+            if !notakeover {
+                mutBuffer = unpad(buffer:buffer)
+            }
+
+            let startTime = Date()
+            
+            let compressed =
+            try mutBuffer.compressStream(with: compressorBox.value!,
+                                         flush: .sync,
+                                         allocator: channel.allocator)
+                                        //rfc says to strip padding word off compressed
+            mutBuffer = unpad(buffer: mutBuffer)
+            if logging {
+                let endTime = Date()
+                let endSize = compressed.readableBytes
+                
+                logger.trace("compressed \(startSize) to \(endSize) bytes @ \(startSize / endSize) ratio from")
+                switch extendedSocketType {
+                case .server:
+                    logger.trace(" \(serverConfig.deflateConfig.zlibConfig)")
+                case .client:
+                    logger.trace(" \(clientConfig.deflateConfig.zlibConfig)")
+                }
+                
+                logger.trace("in \(startTime.distance(to: endTime))")
+            }
+            
+            if notakeover {
+                logger.trace("resetting compressor stream")
+                try compressorBox.value?.resetStream()
+            }else {
+                logger.trace("not resetting compressor stream.")
+            }
+            
+            var frame = WebSocketFrame(
+                fin: true,
+                opcode: opCode,
+                maskKey: self.makeMaskKey(),
+                data: compressed
+            )
+            
+            frame.rsv1 = true // denotes compression
+            let slice = compressed.getSlice(at:compressed.readerIndex,
+                                                         length: compressed.readableBytes - 4)
+            frame.data = slice ?? compressed
```

Comment:
Why are you doing this _after_ creating the frame, and why do you think that it's okay to do `?? compressed` here? That just means that if there aren't at least 4 bytes in the buffer, you act like you didn't need to do anything, and it seems to me that such a case is much more likely to be an error that should terminate processing.

Reply:
I'd kinda like to drive that out with tests before removing it . Need to at least examine its logic closely.